### PR TITLE
Add 'list' class to list_of vctrs

### PR DESCRIPTION
**Pull request by jeroen**
This adds the `list` class to `list_of` vctrs. This solves the issue with jsonlite (and possibly other packages implementing S4 list methods):

 - https://github.com/tidyverse/tidyr/issues/750
 - https://github.com/jeroen/jsonlite/issues/292

Perhaps jsonlite is ill-behaved, but I think adding the `list` class to `list_of` makes sense?

### CODE REVIEW

File: R/type-list-of.R [hunk 1]
```diff
@@ -62,7 +62,10 @@ new_list_of <- function(x = list(), ptype = logical(), ..., class = character())
   stopifnot(is.list(x))
   stopifnot(vec_size(ptype) == 0)
 
-  new_vctr(x, ..., ptype = ptype, class = c(class, "vctrs_list_of"))
+  # Append 'list' class after 'vctrs_vctr'
+  out <- new_vctr(x, ..., ptype = ptype, class = c(class, "vctrs_list_of"))
+  class(out) <- c(class(out), "list")
```
Comment (hadley) [inline R/type-list-of.R]: I think you can just inline `list` into the class list — the point is not to preserve the existing class, but to lift the base type (which becomes the implicit class) into the subclass. 

I'm still not sure this is correct but I'm thinking about it.

**Comment by hadley** [comment]
Ok, @lionel- has convinced me by saying that we could treat the presence of the list class as a signal that the object is a vector (not a scalar, like e.g. `lm()`). (@DavisVaughan — I think this is how we'll resolve the unnest problem in a user friendly way)

Can you please simplify the test to remove the dependency on a json? I think testing that `list_of` inherits from `list` should be adequate.

**Comment by DavisVaughan** [comment]
I agree with this! This is what I was trying to say in the unnest issue as well

**Comment by lionel-** [comment]
What is the unnest problem?

**Comment by DavisVaughan** [comment]
https://github.com/tidyverse/tidyr/issues/761#issuecomment-534743783

**Comment by lionel-** [comment]
I like this approach because it gives an easier way of making a class compatible with vctrs. Currently you can declare a list to be a vector by defining a proxy, but then vctrs assumes the class is fully implemented with vctrs primitives and no longer falls back to base methods, eg. to `[[` when slicing. Adding an explicit `"list"` class for compatibility with vctrs should be an easy change and undisruptive in most cases.

File: R/type-list-of.R (hunk 1 above)
Comment (jeroen) [reply R/type-list-of.R]: Such that the resulting `list_of` class will be `c("vctrs_list_of", "list", "vctrs_vctr")` ?

Comment (lionel-) [reply R/type-list-of.R]: I think `"list"` should be last because we can't superclass foreign classes.

Comment (jeroen) [reply R/type-list-of.R]: Right, that's why I implemented it this way. So keep the PR as is?

Comment (hadley) [reply R/type-list-of.R]: I wonder if we need another argument to `new_vctr()`? Like `new_vctr(x, ..., ptype = ptype, class = c(class, "vctrs_list_of"), preserve_implict_class = TRUE)` (but a better name)

Comment (lionel-) [reply R/type-list-of.R]: oh yes that's occasionally useful.

```r
explicit_base_class = TRUE
base_class = TRUE
subclass_base_type = TRUE
```

Maybe the first one?

Comment (hadley) [reply R/type-list-of.R]: Or `explicit_type`? Otherwise, I like `explicit_base_class` best of the options so far.

Comment (hadley) [reply R/type-list-of.R]: Another option would be `include_type_in_class` (that's too long, but might suggest other ideas)

Comment (DavisVaughan) [reply R/type-list-of.R]: `explicit_typeof` might also work, depending on how we'd extract the base type

Comment (jeroen) [reply R/type-list-of.R]: Should I try to add something like this to the PR? Anyone with strong opinions want to take it from here?

Comment (lionel-) [reply R/type-list-of.R]: No worries @jeroen, I can finish the PR when I get back to work on vctrs.

Comment (hadley) [reply R/type-list-of.R]: How about `extends_type`?

**Pull request closed without merging**